### ISSUE TITLE
Zenject and UNET

### ISSUE THREAD

**Issue by Leonard0SSilva** [ssannandeji/Zenject-2019#187]
Has anyone done a multiplayer game with zenject? I'm having trouble instantiating an object over the network using:
DiContainer or Factory with
NetworkServer.Spawn (myObject) ;.
Theoretically this should return a copy of the binned instance on all clients but returns a copy without a bind.
I also tried instantiating my object and NetworkServer.Spawn (myObject); In the bind instance but I have the same problem.
I'm not sure if I should turn MonoBehaviour's zenject base into NetworkBehaviour, already tried all of them I think (Context and MonoInstallerBase). MonoInstallerBase gives you an error and you can not test it if you try to make it into NetworkBehaviour.
![image](https://cloud.githubusercontent.com/assets/19909277/21618584/c6c4d580-d1d2-11e6-9d53-ce243679513a.png)
Not sure which way to go, thank you for reading.

**Comment by svermeulen** [comment]
I'm not familiar with UNET, but if UNET is handling instantiating the object, you can still use Zenject - you just have to manually inject it once UNET provides it to you.  So, you can add a reference to DiContainer then call `_container.Inject(gameObject)` on the game object that UNET gives you.  Does that address the problem?

**Comment by Leonard0SSilva** [comment]
Still not working
This code is binded in a scene context and runs only once on the server.

     Public virtual void Initialize ()
     {
         Var player = container.InstantiatePrefab (settings.player);
         NetworkServer.Spawn (player.gameObject);
     }

The player instantiated on the server is inject and works normally, I created a class that does what you said but the clones instantiated in the clients do not inject.

public class PlayerResolve : NetworkBehaviour
{
    DiContainer container;

    [Inject]
    void Construct(DiContainer _container)
    {
        container = _container;
    }

    public void Start()
    {
        container.InjectGameObject(gameObject);
    }
}

Please tell me if I do not understand something.

**Comment by svermeulen** [comment]
Is PlayerResolve attached to the clones?  It might be because the [Inject] method of the clones is never called.  You need to call `container.Inject(gameObject)` on the clones from a class that itself has been created through Zenject, so that it has access to the container.  Then you somehow need to detect when clones are created so you can call Inject at that time

**Comment by Sixze** [comment]
You can do this:
* Create this two scripts:
```csharp
using UnityEngine;
using Zenject;

public class PrefabInjectorInstaller : MonoInstaller<PrefabInjectorInstaller>
{
    public static DiContainer SceneContainer { get; private set; }

    public override void InstallBindings();
    {
        SceneContainer = Container;
    }

    public static void InjectGameObject(GameObject gameObject)
    {
        SceneContainer.InjectGameObject(gameObject);
    }
}
```
```csharp
public class PrefabInjector: MonoBehaviour
{
    private void Awake()
    {
        PrefabInjectorInstaller.InjectGameObject(gameObject);
        Destroy(this);
    }
}
```
* Change `PrefabInjector` script execution order to something like this:
![1](https://cloud.githubusercontent.com/assets/16257871/21659882/b9ffb5f2-d2dd-11e6-9bd1-4252a067593b.PNG)
* Attach `PrefabInjector` scrip to prefabs, that will be spawned by server and attach `PrefabInjectorInstaller` script to `SceneContext` game object
* Finally you got something like self-binding prefab, that you can instantiate whatever way you want

Maybe it's not the best way, but it is the easiest due to the fact that on the client side in UNET you can't control the process of instantiating prefabs (call to instantiate hidden deep within private methods) and bindings do not synchronize over network (but you can try to do something similar :)).

I think that it would be nice to have an opportunity like this for self-binding prefab in zenject out of the box, just component (perhaps with some customization), which you simply attach to prefab :)

**Comment by Jamjardavies** [comment]
I've done some networking code while using Zenject.

I don't have any code available currently, but you could try creating your own factory, and for the Create method, do something like this:

**Note**: This is just off my head and might not work exactly.

```
public class NetworkFactory : IFactory<GameObject>
{
    [Inject]
    private DiContainer m_container = null;

    public GameObject Create(GameObject prefab)
    {
        var go = NetworkServer.Spawn(prefab);

        m_container.Inject(go);

        return go;
    }
}
```

Also, the objects you want to spawn, shouldn't be installers. These should be created in the scene ready. If you need bindings only for a player, then you should use facades.

I am hoping to get some time to supply a networking demo with Zenject that could be supplied as part of Zenject.

**Comment by svermeulen** [comment]
@SteamSector The only issue with the self-binding prefab solution is that the prefab might be instantiated in a subcontainer.  The way you have it, every prefab instantiated this way would use the scene container.  This would work fine if you aren't using subcontainers though.
